cheetah/accelerator/quadrupole.py

```diff
@@ -72,6 +93,135 @@ def transfer_map(self, energy: torch.Tensor) -> torch.Tensor:
+            return incoming
+        elif self.tracking_method == "cheetah":
+            if isinstance(incoming, ParameterBeam):
+                tm = self.transfer_map(incoming.energy)
+                mu = torch.matmul(tm, incoming._mu.unsqueeze(-1)).squeeze(-1)
+                cov = torch.matmul(
+                    tm, torch.matmul(incoming._cov, tm.transpose(-2, -1))
+                )
+                return ParameterBeam(
+                    mu,
+                    cov,
+                    incoming.energy,
+                    total_charge=incoming.total_charge,
+                    device=mu.device,
+                    dtype=mu.dtype,
+                )
+            elif isinstance(incoming, ParticleBeam):
+                tm = self.transfer_map(incoming.energy)
+                new_particles = torch.matmul(incoming.particles, tm.transpose(-2, -1))
+                return ParticleBeam(
+                    new_particles,
+                    incoming.energy,
+                    particle_charges=incoming.particle_charges,
+                    device=new_particles.device,
+                    dtype=new_particles.dtype,
+                )
+            else:
+                raise TypeError(
+                    f"Parameter incoming is of invalid type {type(incoming)}"
+                )
+        elif self.tracking_method == "bmadx":
+            if isinstance(incoming, ParameterBeam):
+                raise NotImplementedError(
+                    "ParameterBeam tracking is not supported for bmadx tracking method."
+                )
+            elif isinstance(incoming, ParticleBeam):
+                return self._bmadx_track(incoming)
+            else:
+                raise TypeError(
+                    f"Parameter incoming is of invalid type {type(incoming)}"
+                )
+
+    def _bmadx_track(self, incoming: ParticleBeam) -> ParticleBeam:
+        # compute bmad coords and p0c:
+        mc2 = electron_mass_eV.to(
+            device=incoming.particles.device, dtype=incoming.particles.dtype
+        )
+        bmad_coords, p0c = cheetah_to_bmad_coords(
+            incoming.particles, incoming.energy, mc2
+        )
+        x = bmad_coords[..., 0]
+        px = bmad_coords[..., 1]
+        y = bmad_coords[..., 2]
+        py = bmad_coords[..., 3]
+        z = bmad_coords[..., 4]
+        pz = bmad_coords[..., 5]
+
+        # quad params:
+        length = self.length
+        k1 = self.k1.clone()
+        n_step = self.num_steps
+        step_len = length / n_step
+
+        x_off = self.misalignment[..., 0]
+        y_off = self.misalignment[..., 1]
+        tilt = self.tilt
+
+        b1 = k1 * length
+
+        # Begin bmadx tracking:
+        x, px, y, py = offset_particle_set(x_off, y_off, tilt, x, px, y, py)
+
+        for _i in range(n_step):
+            rel_p = 1 + pz  # Particle's relative momentum (P/P0)
+            k1 = b1 / (length * rel_p)
+
+            tx, dzx = quad_mat2_calc(-k1, step_len, rel_p)
```

Comment:
Maybe rename the `quad_mat2_calc` to something more readable like `calculate_second_order_quadrupole_matrix`?

Reply:
renamed